src/artemis/experiment_plans/optimise_attenuation_plan.py

```diff
@@ -58,35 +86,230 @@ def arm_devices(xspress3mini, zebra):
+    yield from bps.abs_set(attenuator, transmission, group="set_transmission")
+    yield from bps.abs_set(xspress3mini.set_num_images, 1, wait=True)
+    yield from arm_devices(xspress3mini, zebra)
+
+
+def is_deadtime_optimised(
+    deadtime: float,
+    deadtime_threshold: float,
+    transmission: float,
+    upper_transmission_limit: float,
+) -> bool:
+    if deadtime <= deadtime_threshold or transmission == upper_transmission_limit:
+        if transmission == upper_transmission_limit:
+            LOGGER.warning(
+                f"Deadtime {deadtime} is above threshold {deadtime_threshold} at maximum transmission {upper_transmission_limit}. Using maximum transmission\
+                        as optimised value."
+            )
+        return True
+    else:
+        return False
+
+
+def deadtime_optimisation(
+    attenuator: Attenuator,
+    xspress3mini: Xspress3Mini,
+    zebra: Zebra,
+    transmission: float,
+    upper_transmission_limit: float,
+    lower_transmission_limit: float,
+    increment: float,
+    deadtime_threshold: float,
+    max_cycles: int,
+):
+    """Optimises the attenuation for the Xspress3Mini based on the detector deadtime
+
+    Deadtime is the time after each event during which the detector cannot record another event. This loop adjusts the transmission of the attenuator
+    and checks the deadtime until the deadtime is below the accepted threshold. To protect the sample, the transmission has a maximum value
+
+    Args:
+        attenuator: (Attenuator) Ophyd device
+
+        xspress3mini: (Xspress3Mini) ophyd device
+
+        zebra: (Zebra) Ophyd device
+
+        transmission: (float)
+        The intial transmission value to use for the optimising
+
+        increment: (float)
+        The factor to increase / decrease the transmission each cycle
+
+        deadtime_threshold: (float)
+        The maximum acceptable percentage deadtime
+
+        max_cycles: (int)
+        The maximum number of iterations before an error is thrown
+
+    Returns:
+        optimised_transmission: (float)
+        The final transmission value which produces an acceptable deadtime
+    """
+
+    direction = Direction.POSITIVE
+    LOGGER.info(f"Target deadtime is {deadtime_threshold}")
+
+    for cycle in range(0, max_cycles):
+        yield from do_device_optimise_iteration(
+            attenuator, zebra, xspress3mini, transmission
+        )
+
+        total_time = xspress3mini.channel_1.total_time.get()
+        reset_ticks = xspress3mini.channel_1.reset_ticks.get()
+
+        LOGGER.info(f"Current total time = {total_time}")
+        LOGGER.info(f"Current reset ticks = {reset_ticks}")
+        deadtime = 0
+
+        """ Deadtime is the time after each event during which the detector cannot record another event.
+            The reset ticks PV stops ticking while the detector is unable to process events, so the absolute difference between the total time and the
```

Comment:
I'm not 100% sure about this description of the reset ticks, but given the formula and definition for deadtime it looks like it must be something like this

Reply:
I don't think we should have comments in that we're not sure about. Let's look more inhttps://github.com/DiamondLightSource/python-artemis/issues/776